xskillscore/core/deterministic.py

```diff
@@ -6,14 +6,61 @@
 __all__ = ['pearson_r', 'pearson_r_p_value', 'rmse', 'mse', 'mae']
 
 
-def _preprocess(dim):
+def _preprocess_dims(dim):
+    """Preprocesses dimensions to prep for stacking.
+    
+    Parameters
+    ----------
+    dim : str, list
+        The dimension(s) to apply the function along.
+    """
     if isinstance(dim, str):
         dim = [dim]
     axis = tuple(range(-1, -len(dim) - 1, -1))
     return dim, axis
 
 
-def pearson_r(a, b, dim):
+def _preprocess_weights(a, dim, new_dim, weights):
+    """Preprocesses weights array to prepare for numpy computation.
+    
+    Parameters
+    ----------
+    a : Dataset, DataArray, GroupBy, Variable, numpy/dask arrays or scalars
+        One of the arrays over which the function will be applied.
+    dim : str, list
+        The original dimension(s) to apply the function along.
+    new_dim : str
+        The newly named dimension after running ``_preprocess_dims``
+    weights : Dataset, DataArray, GroupBy, Variable, numpy/dask arrays or scalars
+        Weights to apply to function, matching the dimension size of ``new_dim``.
+    """
+    if weights is None:
+        weights = xr.ones_like(a)  # Evenly weight
+    else:
+        # Scale weights to vary from 0 to 1.
+        weights = weights / weights.max()
+        # Throw error if there are negative weights.
+        if weights.min() < 0:
+            raise ValueError(
+                f"Weights has a minimum below 0. Please submit a weights array "
+                f"of positive numbers."
+                )
+        # Check that the weights array has the same size
+        # dimension(s) as those being applied over.
+        drop_dims = [i for i in a.dims if i not in new_dim]
+        drop_dims = {k: 0 for k in drop_dims}
+        if dict(weights.sizes) != dict(a.isel(drop_dims).sizes):
```

Comment:
This becomes more problematic. I'm just checking to see if dims + sizes align. This might be harder to babysit with ndarrays. I guess this comes down to design philosophy following https://github.com/raybellwaves/xskillscore/issues/29.